README.md

```diff
@@ -132,12 +132,43 @@ currently, there are verification tasks for the ILP32 (32-bit) and the LP64 (64-
 
 In order to obtain verification tasks from the programs and specifications in the repository,
 a simple task-definition mechanism is used.
-For each program, the repository contains a .yml file that specifies 
-  - the subject files (the parts that a program consists of),
-  - the results (for each specification of the program, the expected outcome is indicated), and
-  - the parameters to be used for verification.
+For each program, the repository contains a .yml file that specifies the following items:
+  - `format_version`: the version of the format (a version string)
+  - `input_files`: the subject program files or directories
+    (a file or directory name, or a list of files or directory names, that the program consists of)
+  - `properties`: the properties that constitute the specification of the program,
+    each consisting of the following items:
+    - `property_file`: file that contains a property definition
+    - `expected_verdict`: the intended verification result (`true` or `false`)
+    - `subproperty` (optional): a subproperty of the property that is violated
+      in cases where the property is a conjunction of subproperties (for verdict `false`)
+  - `options`: parameters that are relevant for verification or give extra information:
+    - `language`: programming language that the program is written in (`C` or `Java`)
+    - `data_model` data model of the computer architecture
+      (`ILP32`, `LP64`, see http://www.unix.org/whitepapers/64bit.html)
   
-The [SV-COMP 2019 report] contains a description of the format with an example in Sect. 4 and Fig. 3.
+The [SV-COMP 2019 report] has documented the first version of the repository's task-definition format 1.0,
+and contains a description of the format with an example in Sect. 4 and Fig. 3.
+Format 1.1 adds the `options` dictionary.
```

Comment:
What about backwards compatibility? We could define that for tasks that still have format version 1.0 the language is assumed to be C and the data model ... well that is not so easy to define. Unless we add a special data model `unspecified` or so into the format version 1.1. Then we could say the old tasks have data model `undefined`, and the tool-info modules can then deal with this, e.g. not add additional parameters.

Reply:
Sorry, I missed to read the comment above (https://github.com/sosy-lab/sv-benchmarks/pull/1113#issuecomment-694973380). So if we do not want a default language, which makes sense since we also have Java. Then my proposal would be to just also set the default to the signalling value `unspecified` also for the language option in case someone throws a task definition with the old format at the tools.

Reply:
Why do you assume/want default values? If an entry is missing from (i.e., not defined in) the task definition, I feel the easiest solution is to handle it as undefined.

If there's default values, this allows implicit definitions. This makes reading the task definitions more complicated, because one has to know about the default values (and/or look them up).

Because of this, I would also not set a default of `unspecified` per entry, but rather say that not defining a key means that it's undefined (for all entries, by default).

Reply:
Yeah that would probably also work I guess.

It really depends on how this is implemented in the `tool-info` modules. It is more error-prone to silently pass no value for these options than to explicitly pass a special value that signals the tool-info module developer that they need to deal with that case. But this are high-level considerations, depend on future work and we could still at any time change the implementation of the `tool-info` modules to do this.

How do I specify a task in format 2.0 that is actually equivalent to a task in 1.0? Do I just leave out/empty the `options` or do I add a signalling value (`unspecified` or so).

In the end this also leads to the question whether the `options` in the new format version are mandatory or can be left out? Currently if I read this file then `options.{language,data_model}` need to always be explicitly given, so what would I input there in this case where I want to describe a version 1.0 task with the new format?

Reply:
> It really depends on how this is implemented in the tool-info modules. It is more error-prone to silently pass no value for these options than to explicitly pass a special value that signals the tool-info module developer that they need to deal with that case.

I agree with that. But it's a different question, isn't it? It is about the implementation of 'undefined' in BenchExec. So I wouldn't be concerned with it when defining the format.

> How do I specify a task in format 2.0 that is actually equivalent to a task in 1.0? Do I just leave out/empty the options or do I add a signalling value (unspecified or so).

(you probably mean task in format `1.1`, not `2.0`)
In my understanding, task definitions of version 1.0 were incomplete, because we stil needed information from the `.cfg` files to determine the data model of tasks, and had to look at the program-file suffix to see whether a file is a `C` or `Java` file.
Task-format 1.1 aims to make the task definitions complete, so if you want to define a task in a 1.0-manner, you also just leave some information undefined.

> In the end this also leads to the question whether the options in the new format version are mandatory or can be left out?

If I remember correctly, they are optional. I think @dbeyer mentioned in one discussion that if you don't care about the data model a task is analyzed for, you can leave the item undefined.
This should be added to the documentation.

Reply:
From BenchExec's point of view, every task-definition file that is valid as version 1.0 is also valid if only its version is changed to 1.1. I would suggest the same for tool-info modules, i.e., they should simply treat absence of any part of `options` as before. Otherwise one cannot benchmark older versions of SV-Benchmarks with a newer BenchExec.

Reply:
> I agree with that. But it's a different question, isn't it? It is about the implementation of 'undefined' in BenchExec. So I wouldn't be concerned with it when defining the format.

Yes, I wanted to say exactly that, it is a different question, this became clear to me now :sweat_smile:.

> (you probably mean task in format 1.1, not 2.0)

Yeah my mistake, but actually it should be 2.0. It is not like it is compatible or anything. But of course that is just my opinion, in the end it doesn't matter too much.

> From BenchExec's point of view, every task-definition file that is valid as version 1.0 is also valid if only its version is changed to 1.1

Oh well then I have to correct my statement above and then it makes absolute sense to bump the version to 1.1.

> they should simply treat absence of any part of options as before. Otherwise one cannot benchmark older versions of SV-Benchmarks with a newer BenchExec.

That was exactly what I was aiming at, the options should not be mandatory then :+1:

Reply:
- There is no backwards compatibility regarding task-definition files. (Old: data model in .cfg file, new: data model in .yml file)
- Format 1.1 requires the options to be set. (This is not checked by BenchExec, but by the sv-benchmarks CI, if we would like to check it.)
- Yes, I agree: the format should be 2.0, to emphasize the missing compatibility.
- The discussion above heavily mixes BenchExec issues into this sv-benchmarks issue.

Reply:
2.0: The change is indeed a large change, thus, the 2.0 is fully justified.  It is a major change (breaking potentially lots of things).

Reply:
Sorry, I'm not sure I understand why this is a breaking change. Is the following reasoning correct?

1. The task-definition format defines `options.data_model` and `options.language` as mandatory fields. Because of this, the change is breaking and version 2.0 is justified.
2. BenchExec supports a (less strict) superset of the task-definition format and does not enforce these mandatory fields.

Reply:
Yes, BenchExec supports not exactly our format, but a less strict format.
But this is not a problem, because we can enforce consistency via CI checks.

In addition, as soon as we remove the .cfg files, new versions of sv-benchmarks will not work with old benchmark-definition files. Vice versa, new benchmark-definition files will not work with old versions of sv-benchmarks.

Reply:
There is no dependency between any benchmark definitions and the `.cfg` files. So far the `.cfg` files have only served as documentation and for the CI scripts, but not for the benchmark execution.

As soon as we give up the rule that all tasks in a category have the same data model (which is possible after this), then old benchmark definitions will need adjustment, just as usual when we change categories.